### ISSUE TITLE
RE: Ghosting Issue

### ISSUE THREAD

**Issue by jagadeeshpinter** [2dom/PxMatrix#158]
I 'm using NOde (ESP-12E) module  and arudino Program..I have Ghosting Issue on LED display

Below code reference..

`#include <Ticker.h>
#include <PxMatrix.h>
#include <string.h>

#ifdef ESP8266

Ticker display_ticker;
#define P_LAT 16          /*Define A,B,C,LATCH,OUTPUT ENABLE PINS*/
#define P_A 5
#define P_B 4
#define P_C 12 
#define P_OE 15

#endif

PxMATRIX display(64,32,P_LAT, P_OE,P_A,P_B,P_C);      /*Create Matrix Display Object*/
#ifdef ESP8266
void display_updater()                                /*Display refresh rate - do not change*/
{
  display.display(50);
}
#endif

void display_update_enable(bool is_enable)
{

#ifdef ESP8266                                        /*Display latch time - do not change*/
  if (is_enable)
    display_ticker.attach(0.001, display_updater);
  else
    display_ticker.detach();
#endif
}
//uint16_t Red = display.color565(0, 0, 255);           /*Display Color Definitions*/
uint16_t Red = display.color565(255, 0, 0);
uint16_t Green = display.color565(0, 255, 0);
uint16_t Yellow = display.color565(0, 69,255);
//uint16_t Blue = display.color565(255, 0, 0);
uint16_t Blue = display.color565(0, 0, 255); 
uint16_t Black = display.color565(0, 0, 0);
uint16_t White = display.color565(255,255,255);

void COLOR(char c1){                                 /*Display color switcher function*/
 if(c1=='0')display.setTextColor(Red);
 else if(c1=='1')display.setTextColor(Green);
 else if(c1=='2')display.setTextColor(Yellow);
 else display.setTextColor(Blue);
  }

int xi;short int garb=7;                             /*Operating variables*/
int ti;int key=1;int saavi=0;int thiravu=0;
int itr=0;int buff_ptr=0;
char r1i,r1[25],r2[25],r3[25],c1[25],c2[25],c3[25],buff[300];
int sthas,endhas;
char welcm[9]=" Welcome";

//#################################CLEAR BUFFER#####################################
void clear_buffer(){
        for(int i=0;i<300;i++)buff[i]=0;
        r1i=0;
        for(int i=0;i<25;i++)r1[i]=0;
        for(int i=0;i<25;i++)c1[i]=0;
        for(int i=0;i<25;i++)r2[i]=0;
        for(int i=0;i<25;i++)c2[i]=0;
        for(int i=0;i<25;i++)r3[i]=0;
        for(int i=0;i<25;i++)c3[i]=0;
  }
//#################################GET UART DATA#####################################
void get_uart()
{        if(Serial.available())
        {   buff[itr]=Serial.read();
            itr++;
        }
}
 
//#################################SORT UART DATA#####################################
void sort_uart()
{
  int i,n,p,q,r,s,t;
  
  r1i=buff[0];
  i=1,n=0;
  while(1){
  r1[n]=buff[i];
  i++;n++;
  if(buff[i]=='\0'){r1[n++]='\0';i++;break;}
  }
  
  p=0;
  while(1){
  c1[p]=buff[i];
  i++;p++;
  if(buff[i]=='\0'){c1[p++]='\0';i++;break;}
  }

  q=0;
  while(1){
  r2[q]=buff[i];
  i++;q++;
  if(buff[i]=='\0'){r2[q++]='\0';i++;break;}
  }
  r=0;
  while(1){
  c2[r]=buff[i];
  i++;r++;
  if(buff[i]=='\0'){c2[r++]='\0';i++;break;}
  }
  
  s=0;
  while(1){
  r3[s]=buff[i];
  i++;s++;
  if(buff[i]=='\0'){r3[s++]='\0';i++;break;}
  }
  
  t=0;
  while(1){
  c3[t]=buff[i];
  i++;t++;
  if(buff[i]=='\0'){c3[t++]='\0';i++;break;}
  }
}

void print_uart(){
  Serial.write("r1i =");Serial.write(r1i);Serial.println("\n\r");

Serial.write("r1 =");Serial.println(r1);Serial.println("\n\r");

Serial.write("c1 =");Serial.println(c1);Serial.println("\n\r");

Serial.write("r2 =");Serial.println(r2);Serial.println("\n\r");

Serial.write("c2 =");Serial.println(c2);Serial.println("\n\r");

Serial.write("r3 =");Serial.println(r3);Serial.println("\n\r");
     
Serial.write("c3 =");Serial.println(c3);Serial.println("\n\r");

Serial.write("Full =");Serial.println(buff);Serial.println("\n\r");          

  }
//#################################REMOVE ZEROS#####################################
void remove_lead_zeros()
{
  if(r2[3]=='0' && r2[2]=='0' && r2[1]=='0'){r2[2]=' ';r2[1]=' ';}
  else if(r2[3]!='0' && r2[2]=='0' && r2[1]=='0'){r2[2]=' ';r2[1]=' ';}
  else if(r2[3]=='0' && r2[2]!='0' && r2[1]=='0'){r2[1]=' ';}
  else if(r2[3]!='0' && r2[2]!='0' && r2[1]=='0'){r2[1]=' ';}

  if(r2[6]=='0' && r2[5]=='0' && r2[4]=='0'){r2[5]=' ';r2[4]=' ';}
  else if(r2[6]!='0' && r2[5]=='0' && r2[4]=='0'){r2[5]=' ';r2[4]=' ';}
  else if(r2[6]=='0' && r2[5]!='0' && r2[4]=='0'){r2[4]=' ';}
  else if(r2[6]!='0' && r2[5]!='0' && r2[4]=='0'){r2[4]=' ';}

  if(r2[9]=='0' && r2[8]=='0' && r2[7]=='0'){r2[8]=' ';r2[7]=' ';}
  else if(r2[9]!='0' && r2[8]=='0' && r2[7]=='0'){r2[8]=' ';r2[7]=' ';}
  else if(r2[9]=='0' && r2[8]!='0' && r2[7]=='0'){r2[7]=' ';}
  else if(r2[9]!='0' && r2[8]!='0' && r2[7]=='0'){r2[7]=' ';}
  
  if(r3[3]=='0' && r3[2]=='0' && r3[1]=='0'){r3[2]=' ';r3[1]=' ';}
  else if(r3[3]!='0' && r3[2]=='0' && r3[1]=='0'){r3[2]=' ';r3[1]=' ';}
  else if(r3[3]=='0' && r3[2]!='0' && r3[1]=='0'){r3[1]=' ';}
  else if(r3[3]!='0' && r3[2]!='0' && r3[1]=='0'){r3[1]=' ';}
  
  if(r3[6]=='0' && r3[5]=='0' && r3[4]=='0'){r3[5]=' ';r3[4]=' ';}
  else if(r3[6]!='0' && r3[5]=='0' && r3[4]=='0'){r3[5]=' ';r3[4]=' ';}
  else if(r3[6]=='0' && r3[5]!='0' && r3[4]=='0'){r3[4]=' ';}
  else if(r3[6]!='0' && r3[5]!='0' && r3[4]=='0'){r3[4]=' ';}

  if(r3[9]=='0' && r3[8]=='0' && r3[7]=='0'){r3[8]=' ';r3[7]=' ';}
  else if(r3[9]!='0' && r3[8]=='0' && r3[7]=='0'){r3[8]=' ';r3[7]=' ';}
  else if(r3[9]=='0' && r3[8]!='0' && r3[7]=='0'){r3[7]=' ';}
  else if(r3[9]!='0' && r3[8]!='0' && r3[7]=='0'){r3[7]=' ';}
}

//#################################GENERIC PRINT#####################################

void generic_print(char row1ind,char row1[10],char row2[25],char row3[25], char color1[10],char color2[25],char color3[25])
{
  display.setBrightness(140);                          //######CHANGE DISPLAY BRIGHTNESS -> (For LOW set 70 For MID set 140 For HIGH set 250)#####
  uint16_t text_length1 = strlen(row1);
  uint16_t text_length2 = strlen(row2);
  uint16_t text_length3 = strlen(row3);
  uint16_t lenmax=(text_length2>text_length3?text_length2:text_length3);

  if(row1ind=='$')                                  //STANDARD FORMAT ONLY STATIC
  {
        remove_lead_zeros();
        display.clearDisplay();
        xi=0;ti=0;
        for(int x=0;x<text_length1;x++){
        display.setCursor(xi+ti,0);
        COLOR(color1[x]);
        display.print(row1[x]);
        if(x==0)ti=1;
        if(x==3)ti=2;
        if(x==6)ti=3;
        xi=xi+6;
        //if(digitalRead(10)==HIGH)goto TOP;
          }
        xi=0;ti=0;
        for(int x=0;x<text_length2;x++){
          display.setCursor(xi+ti,11);
          COLOR(color2[x]);
        display.print(row2[x]);
        if(x==0)ti=1;
        if(x==3)ti=2;
        if(x==6)ti=3;
        xi=xi+6;
        //if(digitalRead(10)==HIGH)goto TOP;
          }
          
        xi=0;ti=0;
        for(int x=0;x<text_length3;x++){
          display.setCursor(xi+ti,22);
          COLOR(color3[x]);
        display.print(row3[x]);
        if(x==0)ti=1;
        if(x==3)ti=2;
        if(x==6)ti=3;
        xi=xi+6;
        //if(digitalRead(10)==HIGH)goto TOP;
          }
    }
    else if((row1ind=='#')||(row1ind=='%'))                                   //% data and 2 Rows only
    {
      display.clearDisplay();
        xi=0;ti=0;
        for(int x=0;x<text_length1;x++){
        display.setCursor(xi+ti,0);
        COLOR(color1[x]);
        display.print(row1[x]);
        if(x==0)ti=1;
        if(x==4)ti=4;
        xi=xi+6;
        //if(digitalRead(10)==HIGH)goto TOP;
          }
        xi=7;ti=0;
        for(int x=0;x<text_length2;x++){
          display.setCursor(xi+ti,11);
          COLOR(color2[x]);
        display.print(row2[x]);
        if(x==3)ti=2;
        xi=xi+6;
        //if(digitalRead(10)==HIGH)goto TOP;
          }
          
        xi=7;ti=0;
        for(int x=0;x<text_length3;x++){
          display.setCursor(xi+ti,22);
          COLOR(color3[x]);
        display.print(row3[x]);
        if(x==3)ti=2;
        xi=xi+6;
        //if(digitalRead(10)==HIGH)goto TOP;
          }
      }
      else if(row1ind==',')                                   //ROW 2 and 3 Scroll
      {
            display.clearDisplay();
          display.setTextWrap(false);  
          display.setTextSize(1);
        if(lenmax>11)
        {
          int xpos2=0;
          int i=10;
       
          for (;i>-((xpos2-14+lenmax*6)-62);i--)
          {
            display.clearDisplay();
            xi=0;ti=0;             
            for(int x=1;x<text_length2;x++){
            COLOR(color2[x]);
            display.setCursor(xi+i,11);
            display.print(row2[x]);
            xi=xi+6;
          //  if(digitalRead(10)==HIGH)goto TOP;
          }
            xi=0;ti=0;                                                            //Row1 Static
            for(int x=1;x<text_length3;x++){
            COLOR(color3[x]);
            display.setCursor(xi+i,22);
            display.print(row3[x]);
            xi=xi+6;
            //if(digitalRead(10)==HIGH)goto TOP;
          }
    
                                                    int16_t xp=0;int16_t xp1=0;                                     //Black Static L and R
                                                    for(;(xp<2+6 && xp1<2+6);xp++ && xp1++){
                                                    int16_t yp=11;int16_t yp1=22;
                                                    for(;(yp<11+8&&yp1<22+8);yp++ && yp1++){
                                                    display.drawPixel(xp-1,yp,Black);
                                                    display.drawPixel(xp1,yp1,Black);}}//if(digitalRead(10)==HIGH)goto TOP;}if(digitalRead(10)==HIGH)goto TOP;}

          xi=0;ti=0;                                                            //Row1 Static
          for(int x=0;x<text_length1;x++){
          display.setCursor(xi+ti,0);
          COLOR(color1[x]);
          display.print(row1[x]);
         // if(x==0)ti=1;
          xi=xi+6;
          //if(digitalRead(10)==HIGH)goto TOP;
          }  
          display.setCursor(0,11);                                            //Static L
          COLOR(color2[0]);
          display.print(row2[0]);
          display.setCursor(0,22);                                            //Static R
          COLOR(color3[0]);
          display.print(row3[0]);
          delay(95);
          yield();
          //if(digitalRead(10)==HIGH)goto TOP;
        }     
        //display.clearDisplay();
      }
        
      if(lenmax<=10)                                       //Static all Rows
      {
        display.clearDisplay();
        xi=0;ti=0;                                       //Static Row1
        for(int x=0;x<text_length1;x++){
        display.setCursor(xi+ti,0);
        COLOR(color1[x]);
        display.print(row1[x]);
        //if(xi==1)ti=2;
        xi=xi+6;
          }
        xi=0;ti=0;                                       //Static Row2
        for(int x=0;x<text_length2;x++){
          display.setCursor(xi+ti,11);
          COLOR(color2[x]);
        display.print(row2[x]);
        //if(xi==1)ti=2;
        xi=xi+6;
          }
          
        xi=0;ti=0;                                       //Static Row3
        for(int x=0;x<text_length3;x++){
          display.setCursor(xi+ti,22);
          COLOR(color3[x]);
        display.print(row3[x]);
        //if(xi==1)ti=2;
        xi=xi+6;
          }
      }
   }
}

void welcome_screen(char rw1[9])
{
     display.setCursor(0,0);      
    display.setTextColor(Red);
    display.println(rw1);
    }
  
/*##########INITIALIZE###########/*/
void setup() {
  pinMode(10,INPUT);
  Serial.begin(9600);
  display.begin(8);
  display.setPanelsWidth(2);
  display.setScanPattern(LINE);
  display.setFastUpdate(true);
  display.clearDisplay();
  display_update_enable(true);
  display.setBrightness(140);               //######CHANGE DISPLAY BRIGHTNESS -> (For LOW set 70 For MID set 140 For HIGH set 250)#####
  welcome_screen(welcm);
}

//##########SUPERLOOP###########

void loop() {
  if(digitalRead(10)==LOW)
  {
    get_uart();
    saavi=1;
    thiravu=1;
  }
  else
  {
    if(saavi==1)
    {
      sort_uart();
      while(Serial.available())
        Serial.read();
      saavi=0;
      }
    if(thiravu==1){
      generic_print(r1i,r1,r2,r3,c1,c2,c3);
      clear_buffer();
      thiravu=0;
      itr=0; 
    }
    }
}`

**Comment by 2dom** [comment]
Did you red the readme and tried all remedies for ghosting?

**Comment by jagadeeshpinter** [comment]
Yes we tried ground un used pins,but still problem is there.We checked with multimeter ,it carries .33amp on some led blinks.

**Comment by stephenkamin** [comment]
My ghosting issue was wrong setting for....
display.begin(8);
{4,8,16,32} is the scanning layout options.

**Comment by 2dom** [comment]
OK ... closing
